internal/queryexecutor/executor.go

```diff
@@ -0,0 +1,160 @@
+	marshledIndexDef, _, err := e.db.GetIndexDefinition(dbName)
+	if err != nil {
+		return nil, err
+	}
+
+	if marshledIndexDef == nil {
+		return nil, errors.New("no index has been defined on the database " + dbName)
+	}
+
+	indexDef := map[string]types.Type{}
+	if err := json.Unmarshal(marshledIndexDef, &indexDef); err != nil {
+		return nil, err
+	}
+
+	queryConditions := make(attributesConditions)
+	for attr, c := range conditions {
+		if _, ok := indexDef[attr]; !ok {
+			return nil, errors.New("attribute [" + attr + "] given in the query condition is not indexed")
+		}
+
+		cond, ok := c.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("query syntax error near the attribute [" + attr + "]")
+		}
+
+		attrType := indexDef[attr]
+		conds := &attrConditions{
+			valueType:  attrType,
+			conditions: make(map[string]interface{}),
+		}
+
+		for opt, v := range cond {
+			if !isValidLogicalOperator(opt) {
+				return nil, errors.New("invalid logical operator [" + opt + "] provided for the attribute [" + attr + "]")
+			}
+
+			if err := validateType(v, attrType); err != nil {
+				return nil, errors.WithMessage(err, "attribute ["+attr+"] is indexed but the value type provided in the query does not match the actual indexed type")
+			}
+
+			if attrType == types.Type_NUMBER {
+				v, err = v.(json.Number).Int64()
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			conds.conditions[opt] = v
+		}
+
+		if err := validateAttrConditions(conds.conditions); err != nil {
+			return nil, errors.WithMessage(err, "query syntax error near attribute ["+attr+"]")
+		}
+		queryConditions[attr] = conds
+	}
+
+	return queryConditions, nil
+}
+
+func isValidLogicalOperator(opt string) bool {
+	switch opt {
+	case constants.Equal,
+		constants.NotEqual,
+		constants.GreaterThan,
+		constants.LesserThan,
+		constants.GreaterThanOrEqual,
+		constants.LesserThanOrEqual:
+		return true
+	default:
+		return false
+	}
+}
+
+func validateType(v interface{}, t types.Type) error {
+	kind := reflect.TypeOf(v).Kind()
+	switch kind {
+	case reflect.String:
+		isNumber := reflect.TypeOf(v).Name() == "Number"
+
+		if t == types.Type_STRING && !isNumber {
+			return nil
+		} else if t == types.Type_NUMBER && isNumber {
+			return nil
+		} else {
+			providedType := kind.String()
+			if isNumber {
+				providedType = "number"
+			}
+			return errors.New("the actual type [" + strings.ToLower(t.String()) + "]" +
+				" does not match the provided type [" + providedType + "]")
+		}
+
+	case reflect.Bool:
+		if t == types.Type_BOOLEAN {
+			return nil
+		}
+		return errors.New("the actual type [" + strings.ToLower(t.String()) + "]" +
+			" does not match the provided type [" + kind.String() + "]")
+	default:
+		return errors.New("the actual type [" + strings.ToLower(t.String()) + "]" +
+			" does not match the provided type [" + kind.String() + "]")
+	}
+}
+
+func validateAttrConditions(conds map[string]interface{}) error {
```

Comment:
could you write in words a formal definition of what you expect here?

in addition, for readability, you could also encapsulate  `conds[constants.Equal]interface{}` into its own type and have the validation method on it... 

PS
actually you could do this on the `attrConditions` struct, and then because you know the type, you could also check that the ranges `x<=y` (see comment bellow) make sense...

Reply:
It seems that this condition defines an interval, i.e [x:y] (x:y] [x:y) (x:y) or [x:x]-> equals
What if x>y ? this results in an empty interval... ?

Reply:
Done. Added the rules in the comment. Further, it will be added when we write the documentation. Yes, `x>y` is allowed and it will result in empty result set.